src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2787,3 +2789,115 @@ def find_q(m, m4_fac, D):
                         yield Et
                     except ValueError:
                         pass
+
+def EllipticCurve_with_prime_order(N):
+    r"""
+    Given a prime number ``N``, find another prime number `p` and construct an
+    elliptic curve `E` defined over `\mathbb F_p` such that
+    `\#E(\mathbb F_p) = N`.
+
+    INPUT:
+
+    - ``N`` -- integer; the order for which we seek an elliptic curve. Must be a
+      prime number.
+
+    OUTPUT: an elliptic curve `E/\mathbb F_p` of order ``N``
+
+    EXAMPLES::
+
+        sage: N = next_prime(int(b'sagemath'.hex(), 16))
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E
+        Elliptic Curve defined by y^2 = x^3 + 4757897140353078952*x +
+        1841350074072114366 over Finite Field of size 8314040074357871443
+        sage: E.order() == N
+        True
+
+    The execution time largely depends on the input::
+
+        sage: N = 125577861263605878504082476745517446213
+        sage: E = EllipticCurve_with_prime_order(N) # Takes ~1 second.
+        sage: E.order() == N
+        True
+
+    TESTS::
+
+        sage: N = next_prime(123456789)
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E.order() == N
+        True
+
+        sage: for N in prime_range(3, 100):
+        ....:     E = EllipticCurve_with_prime_order(N)
+        ....:     assert E.order() == N
+
+        sage: N = 123456789
+        sage: E = EllipticCurve_with_prime_order(N)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+        sage: E = EllipticCurve_with_prime_order(0)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+    .. NOTE::
+
+        Depending on the input, this function may run forever.
+
+    ALGORITHM: [BS2007]_, Algorithm 2.2
+    """
+    from sage.arith.misc import is_prime, next_prime
+    from sage.combinat.subset import powerset
+    from sage.functions.other import floor
+    from sage.misc.functional import symbolic_prod as product
+    from sage.quadratic_forms.binary_qf import BinaryQF
+    from sage.schemes.elliptic_curves.cm import hilbert_class_polynomial
+
+    if not is_prime(N):
+        raise ValueError("input order is not prime")
+
+    # The algorithm consists of multiple ‘search rounds’ for a suitable
+    # discriminant `D`, `r` defines the number of rounds. We expect this
+    # algorithm to terminate after a number of rounds that is polynomial in
+    # loglog N.
+    r = 0
+    S = []
+    # First prime. Iterating over the primes by chunks of size log(`N`).
+    p = next_prime(0)
+
+    while True:
+        while p < (r + 1) * floor(N.log()):
```

Comment:
Might be a bit cleaner if you use the `primes` or `prime_range` function here. `prime_range` is faster but uses more memory. `primes` is an iterator so closer to what you're doing here.

Reply:
I think `S = prime_range(3,  ceil((r + 1) * floor(N.log())))` should work (assuming it's supposed to start at 3). Then you can remove the whole while loop.

Reply:
Indeed, this is quite faster to use `prime_range()`.

Original 
```py
sage: %%timeit S = []
....: p = 3
....: r = 0
....: while p < (r + 1) * floor(log(N)):
....:      S.append(p)
....:      p = next_prime(p)
....: 
729 µs ± 32.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
compared to
```py
sage: %timeit S = prime_range(3, ceil((11 + 1) * log(N)))
35.1 µs ± 239 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

And it's interesting to notice that iterating over it and appending to a list is slower (neglecting the list initializqtion of approx ≃9 ns)
```py
sage: %%timeit
....: t = []
....: for p in prime_range(ceil(11 * log(N)), ceil((11 + 1) * log(N))):
....:     t.append(p)
....: 
42.6 µs ± 550 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

I suggest changes in cef063eae9e4a06a31144907dacdcd865c081764

Thanks for the review!